### PR TITLE
Fast sample environment schema

### PR DESCRIPTION
**Pull request by SkyToGround**
So this is probably the first real job for the schemas working group, right? There is a need for a schema that can be used to make fast sample environment data available. This is it. The fields in the schema are the ones that I think are appropriate. There are a few points that might be worth discussing however:

 * **No oversampling factor** I currently have no implementation of decimating the data as it is received by the EFU. Thus it is in principle possible to calculate the oversampling factor by knowing the native sampling frequency of the hardware and using the time delta. However, we might want to actually provide all the relevant information in the schema including more clear information on if decimation/oversampling was used.
 * **Time stamps** If sampling at 44 Msps, the bandwidth requirements for providing the nanosecond time stamp is 4 times that of the actual sample data. Thus I have for now decided that providing time stamps for each and every sample is optional. There is a possibility of having relative time stamps that are an offset to the packet time stamp. However, that is kind of useless in my opinion as you then might as well calculate the time stamps from the delta time instead.
 * **Time stamp location** Is this something we want? I feels like such a waste not knowing the absolute high resolution time stamp of an oversampled sample which is why I implemented this.
 * **Monotonically increasing counter** This seems like a good extra added test for detecting packet/message loss. It might be that we will trust the systems enough that we do not want this.
 
 
 NOTE: This pull request is more or less blocking my work so the sooner you can take a look at it the better.

### CODE REVIEW

File: schemas/senv_data.fbs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+//NOTE: This schema is under development.
+//If you decide to use it, be aware that backwards compatibility is not guaranteed.
+
+//Used to transmit fast sample environment data
+
+file_identifier "senv";
+
+enum Location : byte { Unknown = 0, Start, Middle, End }
+
+table senv_data {
```
Comment (matthew-d-jones) [inline schemas/senv_data.fbs]: We've used CamelCase for tables and snake_case for fields in other schema.
Now we're using LLVM guidelines maybe it makes sense to use CamelCase for fields, but tables should still be CamelCase.

File: schemas/senv_data.fbs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    Name: string (required);     // Name of the device plus channel number (e.g. "SomeName_3")
```
Comment (matthew-d-jones) [inline schemas/senv_data.fbs]: Channel number is already a field, would it make more sense for this field to be `DeviceName` instead and omit the channel number?

File: schemas/senv_data.fbs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    Channel: int;                // Channel number, currently a value from 0 to 3
+    PacketTimeStamp: ulong;      // The time stamp of the first sample in the value vector
```
Comment (matthew-d-jones) [inline schemas/senv_data.fbs]: In what units, measured from what epoch?

File: schemas/senv_data.fbs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    TimeDelta: double;           // Time in nano seconds between samples
```
Comment (matthew-d-jones) [inline schemas/senv_data.fbs]: Why double, is there a chance of requiring greater than nanosecond precision?

File: schemas/senv_data.fbs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    TimeStampLocation: Location; // Relevant only when oversampling has been done; is the time stamp from the start, midddle or end of the samples that were summed to produce each oversampled sample
```
Comment (matthew-d-jones) [inline schemas/senv_data.fbs]: Typo "midddle"

File: schemas/senv_data.fbs [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+enum Location : byte { Unknown = 0, Start, Middle, End }
+
+table senv_data {
+    Name: string (required);     // Name of the device plus channel number (e.g. "SomeName_3")
+    Channel: int;                // Channel number, currently a value from 0 to 3
+    PacketTimeStamp: ulong;      // The time stamp of the first sample in the value vector
+    TimeDelta: double;           // Time in nano seconds between samples
+    TimeStampLocation: Location; // Relevant only when oversampling has been done; is the time stamp from the start, midddle or end of the samples that were summed to produce each oversampled sample 
+    Values: [ushort] (required); // The sample values. Note that only the lower 14bits are used.
+    TimeStamps: [ulong];         // OPTIONAL, time stamps of each individual sample
```
Comment (matthew-d-jones) [inline schemas/senv_data.fbs]: In what units, measured from what epoch?

File: schemas/senv_data.fbs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    MessageCounter: int;         // Monotonically increasing counter
```
Comment (matthew-d-jones) [inline schemas/senv_data.fbs]: `ulong` or possibly `uint` might be more appropriate?
We went for `message_id: ulong` in https://github.com/ess-dmsc/streaming-data-types/blob/master/schemas/ev42_events.fbs but I prefer the name you've used here.

File: schemas/senv_data.fbs (hunk 5 above)
Comment (SkyToGround) [reply schemas/senv_data.fbs]: Oops!

File: schemas/senv_data.fbs (hunk 4 above)
Comment (SkyToGround) [reply schemas/senv_data.fbs]: Having nanosecond resolution is much better than required when only worrying about the time delta between two samples. However, if you want to use the delta to calculate the time stamp of sample number 1000, there is a possibility of being off by 500 ns in some circumstances.

File: schemas/senv_data.fbs (hunk 2 above)
Comment (SkyToGround) [reply schemas/senv_data.fbs]: Having the channel number as an integer is a (very cheap) nicety that I think we should have. As for the name, there are two options in my opinion:
1. Have the name be specific to the ADC device and let the file writer module construct a unique string from  when overriding `FBSchemaReader::sourcename_impl()`.
2. Have the name be unique to the channel.

I obviously went with the choice 2 as this is what the other file writing modules currently do. I do not have a problem with changing that however.

File: schemas/senv_data.fbs (hunk 7 above)
Comment (SkyToGround) [reply schemas/senv_data.fbs]: Doing a quick calculation, the message counter will in the worst case scenario wrap within 20 hours when using a uint. I will change that to ulong.

File: schemas/senv_data.fbs (hunk 6 above)
Comment (SkyToGround) [reply schemas/senv_data.fbs]: See comment above about epoch.

**Comment by SkyToGround** [comment]
@matthew-d-jones , care to comment on the points brought up in the pull request message (first comment)?

**Comment by matthew-d-jones** [comment]
@SkyToGround Thanks for the changes and explanations. Adding an oversampling flag and factor seems worthwhile, after all, there's no overhead for them if they are not set. If there would be several fields for oversampling then consider putting them in their own table.
I think it is sensible to include time stamp location.
If we think we may be losing messages at some point then the counter may prove useful as a debugging tool, even if the production consumer software does not use it. As you say, we may decide later that it is unnecessary, but at that point we can simply choose not to set it (if removing it from the schema is a hassle), in the mean-time it is not a large overhead.

**Comment by SkyToGround** [comment]
Some comments on your comments @mkoennecke :

1. At the moment (and for the foreseeable future as far as I know) we have a specific hardware solution that will serve as the ESS fast sample environment ADC. I see no great obstacle as such to modify the schema in order to support higher resolution ADCs though also allowing uint16 is a requirement in my opinion due to bandwidth limitations (see response 3).

2. This is a good point and one that I had ignored for now due to a lack of information on possible sample parameters, bandwidth limitations and how this should be configured.

3. With the current sampling hardware that we have, the ADC is sampling in lock step with timing system clock. Thus it is impossible to detect jitter on that clock from the ADC. For this reason alone, there is no good reason to provide a time stamp for each sample other than it being nice to have at the receiving end (i.e in the short term only the file writer). Furthermore, the ADC is actually sampling at a rate of 44 million samples per second and as I want to preserve the ability to actually store every one of those samples, if we require that we have a 64 bit timestamp for each sample, we increase the bandwidth requirements (for one ADC channel) from ~100MB/s to ~500MB/s.

4. See above.

5. I honestly do not remember much of the discussion we had on timestamps more than a year ago (?) but the end result was that we decided to use uint64, storing number of nanoseconds since some epoch, as the data type. This is also what is done by almost every other schema in this repository. Do you remember anything interesting about this discussion @matthew-d-jones ?

Seeing as I have a feeling that the schema working group (i.e. @zjttoefs @matthew-d-jones and @mkoennecke ) have not had any discussion on this schema outside of this forum, I want some more comments on the issues brought up by @mkoennecke before I start implementing them.

**Comment by zjttoefs** [comment]
I'm happy with @SkyToGround 's explanations for @mkoennecke 's comments. There may be other message types for ADC values converted to physical quantities. 
The documentation on the epoch that @matthew-d-jones pointed out is not only missing here. We should have a ticket to address that everywhere.

**Comment by zjttoefs** [comment]
That you sneak in changes after the review and then merge right away is a bit odd.

The oversampling factor is of a similar quality as some other low speed configuration parameters, like termination type or voltage range (let alone conversion to a real quantity). Since none of those are part of the high frequency messages, I don't understand how the oversampling is special. What's the idea?

**Comment by SkyToGround** [comment]
Well, it was one of the questions I asked of you (see first comment) and that @matthew-d-jones responded should be there. Seeing as I do not disagree with him and because it is almost no extra development effort, I decided to add it. One of the reasons I was slightly skeptical about implementing conversion to a real quantity right now is because of the development and design work that this requires and seeing as you @zjttoefs and @matthew-d-jones had already accepted the current schema I decided that the conversion could wait for the next version.

Of course, I can easily roll back the merge if you prefer.

**Pull request merged**